src/ray/raylet/pubsub/pubsub_client.cc

```diff
@@ -0,0 +1,161 @@
+  }
+  auto subscription_it = subscription_map_.find(owner_worker_id);
+  RAY_CHECK(subscription_it != subscription_map_.end());
+  RAY_CHECK(
+      subscription_it->second.subscription_callback_map_
+          .emplace(object_id, std::make_pair(std::move(subscription_callback),
+                                             std::move(subscription_failure_callback)))
+          .second);
+
+  // Send a subscription message.
+  auto owner_client = owner_client_pool_.GetOrConnect(owner_address);
+  rpc::SubscribeForObjectEvictionRequest wait_request;
+  wait_request.set_object_id(object_id.Binary());
+  wait_request.set_intended_worker_id(owner_address.worker_id());
+  wait_request.mutable_subscriber_address()->CopyFrom(subscriber_address);
+  owner_client->SubscribeForObjectEviction(
+      wait_request,
+      [this, owner_address, object_id](
+          Status status, const rpc::SubscribeForObjectEvictionReply &reply) {
+        if (!status.ok()) {
+          auto failure_callback_result = GetFailureCallback(owner_address, object_id);
+          auto still_subscribed = failure_callback_result.first;
+          if (still_subscribed) {
+            failure_callback_result.second(object_id);
+            RAY_CHECK(UnsubscribeObject(owner_address, object_id));
+          }
+        }
+      });
+}
+
+bool PubsubClient::UnsubscribeObject(const rpc::Address &owner_address,
+                                     const ObjectID &object_id) {
+  const auto owner_worker_id = WorkerID::FromBinary(owner_address.worker_id());
+  RAY_LOG(DEBUG) << "Unsubscribing an object " << object_id << " from "
+                 << owner_worker_id;
+  auto subscription_it = subscription_map_.find(owner_worker_id);
+  if (subscription_it == subscription_map_.end()) {
+    return false;
+  }
+  auto &subscription_callback_map = subscription_it->second.subscription_callback_map_;
+  auto subscription_callback_it = subscription_callback_map.find(object_id);
+  if (subscription_callback_it == subscription_callback_map.end()) {
+    return false;
+  }
+  subscription_callback_map.erase(subscription_callback_it);
+  return true;
+}
+
+void PubsubClient::MakeLongPollingPubsubConnection(
+    const rpc::Address &owner_address, const rpc::Address &subscriber_address) {
+  const auto owner_worker_id = WorkerID::FromBinary(owner_address.worker_id());
+  RAY_LOG(DEBUG) << "Make a long polling request to " << owner_worker_id;
+
+  auto owner_client = owner_client_pool_.GetOrConnect(owner_address);
+  rpc::PubsubLongPollingRequest long_polling_request;
+  long_polling_request.mutable_subscriber_address()->CopyFrom(subscriber_address);
+
+  owner_client->PubsubLongPolling(
+      long_polling_request, [this, owner_address, subscriber_address](
+                                Status status, const rpc::PubsubLongPollingReply &reply) {
+        HandleLongPollingResponse(owner_address, subscriber_address, status, reply);
```

Comment:
Nit: Instead of a lambda immediately invoking `HandleLongPollingResponse`, you could `std::bind` `HandleLongPollingResponse` to `owner_address` and `subscriber_address`, add two placeholders for the status and reply, and pass that directly as the callback.